core/src/main/java/org/apache/iceberg/io/Writer.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * A writer capable of writing files of a single type (i.e. data/delete) within one spec/partition.
+ */
+public interface Writer<T, R> extends Closeable {
+
+  default void write(Iterable<T> rows) throws IOException {
+    for (T row : rows) {
+      write(row);
+    }
+  }
+
+  void write(T row) throws IOException;
+
+  R result();
```

Comment:
Does `result` need to be called after `close`?

Reply:
Yes, that's ensured in all writer implementations.